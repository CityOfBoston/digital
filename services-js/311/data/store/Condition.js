// @flow

import { computed } from 'mobx';

import type { ServiceAttributeConditionalOp } from '../dao/graphql/types';
import type { ServiceAttributeValuesCondition } from '../types';

import type Question from './Question';

// Class to handle a condition of a dependent picklist or dependent question.
// Combined into a ConditionGroup with "AND" or "OR".
//
// A condition is based on the validated current value of a question (not the
// one that is depending on it) and its "holds" is either true or false.
//
// Uses validated value rather than the cached last set value so that e.g.
// nothing is equal to a hidden questionâ€™s value, regardless of what was filled
// in when it was visible.
export default class Condition {
  question: Question;
  operator: ServiceAttributeConditionalOp;
  value: string | string[] | number;

  constructor(
    { attribute, op, value }: ServiceAttributeValuesCondition,
    questionMap: { [code: string]: Question },
  ) {
    const question = questionMap[attribute];
    if (!question) {
      throw new Error(`Question not found for code: ${attribute}`);
    }

    this.question = question;
    this.operator = op;

    switch (value.type) {
      case 'NUMBER':
        if (value.number == null) {
          throw new Error(
            `Missing number from value: ${JSON.stringify(value)}`,
          );
        }
        this.value = value.number;
        break;

      case 'STRING':
        if (value.string == null) {
          throw new Error(
            `Missing string from value: ${JSON.stringify(value)}`,
          );
        }
        this.value = value.string;
        break;

      case 'STRING_ARRAY':
        if (value.array == null) {
          throw new Error(`Missing array from value: ${JSON.stringify(value)}`);
        }
        this.value = value.array;
        break;

      default:
        throw new Error(`Unknown value type: ${JSON.stringify(value.type)}`);
    }
  }

  @computed
  get holds(): boolean {
    switch (this.operator) {
      case 'eq':
        return this.conditionEq();
      case 'neq':
        return !this.conditionEq();
      case 'in':
        return this.conditionIn();
      case 'gt':
        return this.conditionNumericOp((a, b) => a > b);
      case 'gte':
        return this.conditionNumericOp((a, b) => a >= b);
      case 'lt':
        return this.conditionNumericOp((a, b) => a < b);
      case 'lte':
        return this.conditionNumericOp((a, b) => a <= b);
      default:
        return false;
    }
  }

  conditionEq(): boolean {
    const currentValue = this.question.validatedValue;
    const { value } = this;

    if (typeof value === 'string') {
      return currentValue === value;
    } else if (typeof value === 'number') {
      return parseFloat(currentValue) === value;
    } else if (Array.isArray(value)) {
      if (!Array.isArray(currentValue)) {
        return false;
      } else {
        // Making copies since sort mutates
        const currentValueArr = [...currentValue].sort();
        const expectedValueArr = [...value].sort();

        // array equality!
        return (
          currentValueArr.length === expectedValueArr.length &&
          currentValueArr.filter((el, i) => el === expectedValueArr[i])
            .length === currentValueArr.length
        );
      }
    } else {
      return false;
    }
  }

  conditionIn(): boolean {
    const currentValue = this.question.validatedValue;
    const { value } = this;

    if (!Array.isArray(currentValue) || Array.isArray(value)) {
      return false;
    }

    return currentValue.indexOf(value.toString()) !== -1;
  }

  conditionNumericOp(op: (a: number, b: number) => boolean): boolean {
    const currentValue = this.question.validatedValue;
    const { value } = this;

    if (typeof value !== 'number' || typeof currentValue !== 'string') {
      return false;
    }

    const currentNumber = parseFloat(currentValue);
    return op(currentNumber, value);
  }
}
