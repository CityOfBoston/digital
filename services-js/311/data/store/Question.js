// @flow

import { observable, computed } from 'mobx';
import type { ObservableArray } from 'mobx';

import type { ServiceAttributeDatatype } from '../dao/graphql/types';
import type { ServiceAttribute } from '../types';

import ConditionGroup from './ConditionGroup';

type ValuesArray = Array<{|
  key: string,
  name: string,
|}>;

type ConditionalValuesArray = Array<{|
  conditions: ConditionGroup,
  values: ValuesArray,
|}>;

export default class Question {
  code: string;
  description: string;
  required: boolean;
  type: ServiceAttributeDatatype;
  malformed: boolean;
  _values: ?ValuesArray;
  _conditionalValues: ?ConditionalValuesArray;
  _dependencies: ?ConditionGroup;
  @observable value: ?string | ObservableArray<string>;

  static buildQuestions(attributes: ServiceAttribute[]): Question[] {
    const questionMap = {};
    return attributes.map(attr => {
      const question = new Question(attr, questionMap);
      questionMap[question.code] = question;
      return question;
    });
  }

  constructor(
    attribute: ServiceAttribute,
    questionMap: { [code: string]: Question } = {},
  ) {
    this.code = attribute.code;
    this.description = attribute.description;
    this.required = attribute.required;
    this.malformed = false;
    this._values = attribute.values;
    this._conditionalValues = (attribute.conditionalValues || [])
      .map(({ dependentOn, values }) => ({
        conditions: new ConditionGroup(dependentOn, questionMap),
        values,
      }));
    this._dependencies = attribute.dependencies
      ? new ConditionGroup(attribute.dependencies, questionMap)
      : null;
    this.type = attribute.type;

    switch (this.type) {
      case 'SINGLEVALUELIST':
        this.value = null;
        break;

      case 'MULTIVALUELIST':
        this.value = [];
        break;

      default:
        this.value = '';
        break;
    }
  }

  @computed
  get visible(): boolean {
    return !this._dependencies || this._dependencies.holds;
  }

  @computed
  get requirementsMet(): boolean {
    if (this.malformed) {
      return false;
    }

    const { validatedValue } = this;
    return Array.isArray(validatedValue)
      ? !!validatedValue.length
      : !!validatedValue;
  }

  @computed
  get valueOptions(): ?ValuesArray {
    const { _values: values, _conditionalValues: conditionalValues } = this;

    if (!values || !conditionalValues) {
      return null;
    }

    const activeValuesArr = conditionalValues
      .filter(({ conditions }) => conditions.holds)
      .map(({ values: activeValues }) => activeValues);
    return [...values, ...[].concat(...activeValuesArr)];
  }

  @computed
  get validatedValue(): ?string | string[] {
    if (!this.visible) {
      return null;
    }

    const value = this.value ? this.value.slice() : null;
    const valueOptions = this.valueOptions;

    if (!valueOptions) {
      return value;
    } else if (Array.isArray(value)) {
      return value.filter(v => !!valueOptions.find(({ key }) => key === v));
    } else {
      return valueOptions.find(({ key }) => key === value) ? value : null;
    }
  }
}
