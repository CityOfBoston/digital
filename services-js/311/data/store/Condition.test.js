// @flow

import Condition from './Condition';
import Question from './Question';

const makeStringValue = string => ({
  type: 'STRING',
  string,
  array: null,
  number: null,
});
const makeArrayValue = array => ({
  type: 'STRING_ARRAY',
  string: null,
  array,
  number: null,
});
const makeNumberValue = number => ({
  type: 'NUMBER',
  string: null,
  array: null,
  number,
});

const makeCondition = ({ value, op, currentValue }) => {
  let graphQLValue;
  if (typeof value === 'string') {
    graphQLValue = makeStringValue(value);
  } else if (Array.isArray(value)) {
    graphQLValue = makeArrayValue(value);
  } else if (typeof value === 'number') {
    graphQLValue = makeNumberValue(value);
  } else {
    throw new Error(`Unexpected value type: ${value.toString()}`);
  }

  const question = new Question({
    code: 'code',
    type: 'INFORMATIONAL',
    required: false,
    description: '',
    dependencies: null,
    values: null,
    conditionalValues: null,
  });

  question.value = currentValue;

  return new Condition(
    {
      attribute: 'code',
      op,
      value: graphQLValue,
    },
    { code: question },
  );
};

describe('constructor', () => {
  it('throws if the question canâ€™t be found', () => {
    expect(() => {
      // eslint-disable-next-line no-new
      new Condition(
        { attribute: 'CODE', op: 'eq', value: makeStringValue('value') },
        {},
      );
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('eq', () => {
  test('two strings equal is true', () => {
    expect(
      makeCondition({ value: 'value', op: 'eq', currentValue: 'value' }).holds,
    ).toEqual(true);
  });

  test('two numbers equal is true', () => {
    expect(
      makeCondition({ value: 1, op: 'eq', currentValue: '1' }).holds,
    ).toEqual(true);
  });

  test('two strings not equal is false', () => {
    expect(
      makeCondition({ value: 'value', op: 'eq', currentValue: 'not-value' })
        .holds,
    ).toEqual(false);
  });

  test('two numbers not equal is false', () => {
    expect(
      makeCondition({ value: 1, op: 'eq', currentValue: '2' }).holds,
    ).toEqual(false);
  });

  test('comparing to not a number is false', () => {
    expect(
      makeCondition({ value: 1, op: 'eq', currentValue: 'not-a-number' }).holds,
    ).toEqual(false);
  });

  test('string compared to array is false', () => {
    expect(
      makeCondition({ value: 'value', op: 'eq', currentValue: ['value'] })
        .holds,
    ).toEqual(false);
  });

  test('array compared to string is false', () => {
    expect(
      makeCondition({ value: ['value'], op: 'eq', currentValue: 'value' })
        .holds,
    ).toEqual(false);
  });

  test('subarray compared to array is false', () => {
    expect(
      makeCondition({
        value: ['value-1'],
        op: 'eq',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(false);
  });

  test('arrays in order is true', () => {
    expect(
      makeCondition({
        value: ['value-1', 'value-2'],
        op: 'eq',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(true);
  });

  test('arrays out of order is true', () => {
    expect(
      makeCondition({
        value: ['value-1', 'value-2'],
        op: 'eq',
        currentValue: ['value-2', 'value-1'],
      }).holds,
    ).toEqual(true);
  });

  test('compared to null is false', () => {
    expect(
      makeCondition({ value: 'value', op: 'eq', currentValue: null }).holds,
    ).toEqual(false);
  });
});

describe('neq', () => {
  test('two strings equal is false', () => {
    expect(
      makeCondition({ value: 'value', op: 'neq', currentValue: 'value' }).holds,
    ).toEqual(false);
  });

  test('two not equal strings is true', () => {
    expect(
      makeCondition({ value: 'value', op: 'neq', currentValue: 'not-value' })
        .holds,
    ).toEqual(true);
  });

  test('array not equal to a string is true', () => {
    expect(
      makeCondition({ value: 'value', op: 'neq', currentValue: ['value'] })
        .holds,
    ).toEqual(true);
  });

  test('subarray compared to array is true', () => {
    expect(
      makeCondition({
        value: ['value-1'],
        op: 'neq',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(true);
  });

  test('arrays in order is false', () => {
    expect(
      makeCondition({
        value: ['value-1', 'value-2'],
        op: 'neq',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(false);
  });

  test('arrays out of order is false', () => {
    expect(
      makeCondition({
        value: ['value-1', 'value-2'],
        op: 'neq',
        currentValue: ['value-2', 'value-1'],
      }).holds,
    ).toEqual(false);
  });

  test('compared to null is true', () => {
    expect(
      makeCondition({ value: 'value', op: 'neq', currentValue: null }).holds,
    ).toEqual(true);
  });
});

describe('in', () => {
  test('string in array is true', () => {
    expect(
      makeCondition({
        value: 'value-2',
        op: 'in',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(true);
  });

  test('string not in array is false', () => {
    expect(
      makeCondition({
        value: 'value-3',
        op: 'in',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(false);
  });

  test('array in array is false', () => {
    expect(
      makeCondition({
        value: ['value-2'],
        op: 'in',
        currentValue: ['value-1', 'value-2'],
      }).holds,
    ).toEqual(false);
  });

  test('string in string is false', () => {
    expect(
      makeCondition({ value: 'value', op: 'in', currentValue: 'value' }).holds,
    ).toEqual(false);
  });

  test('number in array is true', () => {
    expect(
      makeCondition({ value: 1, op: 'in', currentValue: ['1', '2'] }).holds,
    ).toEqual(true);
  });

  test('string in null is false', () => {
    expect(
      makeCondition({ value: 'value', op: 'in', currentValue: null }).holds,
    ).toEqual(false);
  });
});

describe('gt', () => {
  test('current value greater than condition is true', () => {
    expect(
      makeCondition({ value: 1, op: 'gt', currentValue: '100' }).holds,
    ).toEqual(true);
  });

  test('current value not greater than condition is false', () => {
    expect(
      makeCondition({ value: 100, op: 'gt', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('string is always false', () => {
    expect(
      makeCondition({ value: '100', op: 'gt', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('greater than null is false', () => {
    expect(
      makeCondition({ value: 1, op: 'gt', currentValue: null }).holds,
    ).toEqual(false);
  });
});

describe('gte', () => {
  test('current value greater than condition is true', () => {
    expect(
      makeCondition({ value: 1, op: 'gte', currentValue: '100' }).holds,
    ).toEqual(true);
  });

  test('current value equal is true', () => {
    expect(
      makeCondition({ value: 100, op: 'gte', currentValue: '100' }).holds,
    ).toEqual(true);
  });

  test('current value not greater than condition is false', () => {
    expect(
      makeCondition({ value: 200, op: 'gte', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('string is always false', () => {
    expect(
      makeCondition({ value: '1', op: 'gte', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('greater than or equal to null is false', () => {
    expect(
      makeCondition({ value: 1, op: 'gte', currentValue: null }).holds,
    ).toEqual(false);
  });
});

describe('lt', () => {
  test('current value less than condition is true', () => {
    expect(
      makeCondition({ value: 101, op: 'lt', currentValue: '100' }).holds,
    ).toEqual(true);
  });

  test('current value not less than condition is false', () => {
    expect(
      makeCondition({ value: 1, op: 'lt', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('string is always false', () => {
    expect(
      makeCondition({ value: '1000', op: 'lt', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('less than null is false', () => {
    expect(
      makeCondition({ value: 1, op: 'lt', currentValue: null }).holds,
    ).toEqual(false);
  });
});

describe('lte', () => {
  test('current value less than condition is true', () => {
    expect(
      makeCondition({ value: 101, op: 'lte', currentValue: '100' }).holds,
    ).toEqual(true);
  });

  test('current value equal is true', () => {
    expect(
      makeCondition({ value: 100, op: 'lte', currentValue: '100' }).holds,
    ).toEqual(true);
  });

  test('current value not less than condition is false', () => {
    expect(
      makeCondition({ value: 1, op: 'lte', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('string is always false', () => {
    expect(
      makeCondition({ value: '1000', op: 'lte', currentValue: '100' }).holds,
    ).toEqual(false);
  });

  test('less than or equal to null is false', () => {
    expect(
      makeCondition({ value: 1, op: 'lte', currentValue: null }).holds,
    ).toEqual(false);
  });
});
