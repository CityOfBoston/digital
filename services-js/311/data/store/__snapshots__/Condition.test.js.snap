// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor throws if the question canâ€™t be found 1`] = `"Question not found for code: CODE"`;
